File: src/main/java/com/epri/dss/shared/impl/HashListImpl.java
package com.epri.dss.shared.impl;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;

import com.epri.dss.common.impl.Utilities;
import com.epri.dss.shared.HashList;

public class HashListImpl implements HashList {

	private class SubList {
		public int nElem;
		public int nAllocated;
		public String[] str;
		public int[] idx;
	}

	private int NumElementsAllocated;

	private int NumLists;

	private int NumElements;

	private String[] StringArray;

	private SubList[] ListArray;

	private int AllocationInc;

	private int LastFind;

	private int LastHash;

	private String LastSearchString;

	protected int InitialAllocation;

	public HashListImpl(int Nelements) {
		super();
		this.NumElements = 0;
		this.InitialAllocation = Nelements;
		this.StringArray = new String[this.NumElements];

		this.NumLists = (int) Math.round(Math.sqrt(Nelements));
		int ElementsPerList = Nelements / NumLists + 1;
		this.AllocationInc = ElementsPerList;
		if (this.NumLists < 1) this.NumLists = 1;  // make sure at least one list
		this.ListArray = new SubList[NumLists];
		for (int i = 0; i < this.NumLists; i++) {
			this.ListArray[i] = new SubList();
			/* Allocate initial sublists to zero; allocated on demand */
			this.ListArray[i].str = new String[0];
			this.ListArray[i].idx = new int[0];
			this.ListArray[i].nAllocated = 0;
			this.ListArray[i].nElem = 0;
		}
		this.NumElementsAllocated = 0;
		this.LastFind = 0;
		this.LastHash = 0;
		this.LastSearchString = "";
	}

	public int getInitialAllocation() {
		return InitialAllocation;
	}

	public void setInitialAllocation(int initialAllocation) {
		InitialAllocation = initialAllocation;
	}

	public int listSize() {
		return NumElements;
	}

	private void resizeSubList(SubList subList) {
		// resize by reasonable amount
		int OldAllocation = subList.nAllocated;
		subList.nAllocated = OldAllocation + AllocationInc;
		subList.str = (String[]) Utilities.resizeArray(subList.str, subList.nAllocated);
		subList.idx = (int[]) Utilities.resizeArray(subList.idx, subList.nAllocated);
	}

	private int hash(String S) {
//		int HashValue = 0;//S.hashCode();
//		for (int i = 0; i < S.length(); i++) {
//			HashValue = ((HashValue << 5) | (HashValue >> 27)) ^ S.charAt(i);
//		}

		int HashValue = S.hashCode();
		return Math.abs(HashValue % NumLists);  // FIXME: negative modulus
	}

	/** Makes the linear string list larger. */
	private void resizeStrArray() {
		NumElementsAllocated += AllocationInc * NumLists;
		StringArray = (String[]) Utilities.resizeArray(StringArray, NumElementsAllocated);
	}

	public int add(String S) {
		int HashNum;
		String SS;

		SS = S.toLowerCase();
		HashNum = hash(SS);

		NumElements += 1;
		if (NumElements > NumElementsAllocated)
			resizeStrArray();

		ListArray[HashNum].nElem += 1;
		if (ListArray[HashNum].nElem > ListArray[HashNum].nAllocated)
			resizeSubList(ListArray[HashNum]);

		// make copy of whole string, lower case
		ListArray[HashNum].str[ListArray[HashNum].nElem - 1] = SS;
		// increments count to string
		StringArray[NumElements - 1] = SS;

		ListArray[HashNum].idx[ListArray[HashNum].nElem - 1] = NumElements - 1;

		return NumElements;
	}

	/**
	 * Repeat find for duplicate string in same hash list.
	 */
	public int find(String S) {

		LastSearchString = S.toLowerCase();
		LastHash = hash(LastSearchString);
		int Result = -1;
		LastFind = -1;

		for (int i = 0; i < ListArray[LastHash].nElem; i++)
			if (LastSearchString.equalsIgnoreCase(ListArray[LastHash].str[i])) {
				Result = ListArray[LastHash].idx[i];
				LastFind = i;
				break;
			}

		return Result;
	}

	/**
	 * Begins search in same list as last.
	 */
	public int findNext() {
		// TODO: Check zero indexing.
		int Result = -1;  // default return
		LastFind += 1;    // start with next item in hash list

		if ((LastHash > 0) && (LastHash <= NumLists)) {
			for (int i = LastFind; i < ListArray[LastHash].nElem; i++)
				if (LastSearchString.equalsIgnoreCase(ListArray[LastHash].str[i])) {
					Result = ListArray[LastHash].idx[i];
					LastFind = i;
					break;
				}
		}

		return Result;
	}

	/**
	 * Makes a linear search and tests each string until a string is found
	 * that matches all the characters entered in s.
	 */
	public int findAbbrev(String S) {
		String Test1, Test2;

		int Result = 0;
		if (S.length() > 0) {
			Test1 = S.toLowerCase();
			for (int i = 0; i < NumElements; i++) {
				Test2 = StringArray[i].substring(0, Test1.length());
				if (Test1.equalsIgnoreCase(Test2)) {
					Result = i;
					break;
				}
			}
		}

		return Result;
	}

	public String get(int i) {
		// TODO: Check zero indexing
		return ((i > 0) && (i <= NumElements)) ? StringArray[i] : "";
	}

	/**
	 * Expands number of elements.
	 *
	 * Creates a new set of string lists and copies the old strings
	 * into the new, hashing for the new number of lists.
	 */
	public void expand(int NewSize) {

		String[] NewStringArray;
		int NewNumLists;
		int ElementsPerList;
		SubList[] NewListArray;
		int HashNum;
		String S;
//		int OldNumLists;

		if (NewSize > NumElementsAllocated) {

//			OldNumLists = NumLists;

			NewStringArray = new String[NewSize];
			NewNumLists = (int) Math.sqrt(NewSize);
			ElementsPerList = NewSize / NewNumLists + 1;
			if (NewNumLists < 1) NewNumLists = 1;  // make sure at least one list
			NewListArray = new SubList[NewNumLists];
			for (int i = 0; i < NumLists; i++) {  // TODO: Check zero indexing.
				/* Allocate initial sublists */
				NewListArray[i].str = new String[ElementsPerList];
				NewListArray[i].idx = new int[ElementsPerList];
				NewListArray[i].nAllocated = ElementsPerList;
				NewListArray[i].nElem = 0;
			}

			NumLists = NewNumLists;  // has to be set so hash function will work

			/* Add elements from old hash list to new hash list */

			for (int i = 0; i < NumElements; i++) {  // TODO: Check zero indexing
				S = StringArray[i];
				HashNum = hash(S);
				NewListArray[HashNum].nElem += 1;
				if (NewListArray[HashNum].nElem > NewListArray[HashNum].nAllocated) {
					resizeSubList(NewListArray[HashNum]);
				}

				NewListArray[HashNum].str[NewListArray[HashNum].nElem] = S;
				NewStringArray[NumElements] = NewListArray[HashNum].str[NewListArray[HashNum].nElem];
				NewListArray[HashNum].idx[NewListArray[HashNum].nElem] = i;
			}

			/* Assign new string and list pointers */

			StringArray = NewStringArray;
			ListArray = NewListArray;
			NumElementsAllocated = NewSize;

		}
	}

	public void dumpToFile(String fname) {
		File FD = new File(fname);
		try {
			PrintWriter F = new PrintWriter(FD);

			F.println(String.format("Number of Hash Lists = %d, Number of Elements = %d", NumLists, NumElements));

			F.println();
			F.println("Hash List Distribution");
			for (int i = 0; i < NumLists; i++)
				F.println(String.format("List = %d, Number of elements = %d", i, ListArray[i].nElem));
			F.println();

			for (int i = 0; i < NumLists; i++) {
				F.println(String.format("List = %d, Number of elements = %d", i, ListArray[i].nElem));
				for (int j = 0; j < ListArray[i].nElem; j++)
					F.println("\"" + ListArray[i].str[j] + "\"  Idx= " + ListArray[i].idx[j]);
				F.println();
			}

			F.println("LINEAR LISTING...");
			for (int i = 0; i < NumElements; i++)
				F.println(i + " = \"" + StringArray[i] + "\"");

			F.close();
		} catch (FileNotFoundException e) {
			// TODO Auto-generated catch block
		}
	}

	public void clear() {
		for (int i = 0; i < NumLists; i++) {
			ListArray[i].nElem = 0;
			for (int j = 0; j < ListArray[i].nAllocated; j++)
				ListArray[i].str[j] = "";
		}

		for (int i = 0; i < NumElementsAllocated; i++)
			StringArray[i] = "";

		NumElements = 0;
	}

}

File: pydss/shared/HashListImpl.py
from dss.shared.HashList import (HashList,)
from dss.common.impl.Utilities import (Utilities,)
# from java.io.File import (File,)
# from java.io.FileNotFoundException import (FileNotFoundException,)
# from java.io.PrintWriter import (PrintWriter,)


class HashListImpl(object, HashList):

    class SubList(object):
        nElem = None
        nAllocated = None
        str = None
        idx = None

    numElementsAllocated = None
    numLists = None
    numElements = None
    stringArray = None
    listArray = None
    allocationInc = None
    lastFind = None
    lastHash = None
    lastSearchString = None
    initialAllocation = None

    def __init__(self, nelements):
        super(HashListImpl, self)()
        self.numElements = 0
        self.initialAllocation = nelements
        self.stringArray = [None] * self.numElements
        self.numLists = self.Math.round(self.Math.sqrt(nelements))
        ElementsPerList = (nelements / self.numLists) + 1
        self.allocationInc = ElementsPerList
        if self.numLists < 1:
            self.numLists = 1
            # make sure at least one list
        self.listArray = [None] * self.numLists
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.numLists):
                break
            self.listArray[i] = self.SubList()
            # Allocate initial sublists to zero; allocated on demand
            self.listArray[i].str = [None] * 0
            self.listArray[i].idx = [None] * 0
            self.listArray[i].nAllocated = 0
            self.listArray[i].nElem = 0
        self.numElementsAllocated = 0
        self.lastFind = 0
        self.lastHash = 0
        self.lastSearchString = ''

    def getInitialAllocation(self):
        return self.initialAllocation

    def setInitialAllocation(self, allocation):
        self.initialAllocation = allocation

    def listSize(self):
        return self.numElements

    def resizeSubList(self, subList):
        # resize by reasonable amount
        oldAllocation = subList.nAllocated
        subList.nAllocated = oldAllocation + self.allocationInc
        subList.str = Utilities.resizeArray(subList.str, subList.nAllocated)
        subList.idx = Utilities.resizeArray(subList.idx, subList.nAllocated)

    def hash(self, s):
        # int HashValue = 0;//S.hashCode();
        # for (int i = 0; i < S.length(); i++) {
        # HashValue = ((HashValue << 5) | (HashValue >> 27)) ^ S.charAt(i);
        # }
        HashValue = s.hashCode()
        return self.Math.abs(HashValue % self.numLists)
        # FIXME: negative modulus

    def resizeStrArray(self):
        """Makes the linear string list larger."""
        self.numElementsAllocated += self.allocationInc * self.numLists
        self.stringArray = Utilities.resizeArray(self.stringArray, self.numElementsAllocated)

    def add(self, s):
        ss = s.toLowerCase()
        hashNum = self.hash(ss)
        self.numElements += 1
        if self.numElements > self.numElementsAllocated:
            self.resizeStrArray()
        self.listArray[hashNum].nElem += 1
        if self.listArray[hashNum].nElem > self.listArray[hashNum].nAllocated:
            self.resizeSubList(self.listArray[hashNum])
            # make copy of whole string, lower case
        self.listArray[hashNum].str[self.listArray[hashNum].nElem - 1] = ss
        # increments count to string
        self.stringArray[self.numElements - 1] = ss
        self.listArray[hashNum].idx[self.listArray[hashNum].nElem - 1] = self.numElements - 1
        return self.numElements

    def find(self, s):
        """Repeat find for duplicate string in same hash list."""
        self.lastSearchString = s.toLowerCase()
        self.lastHash = self.hash(self.lastSearchString)
        result = -1
        self.lastFind = -1
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.listArray[self.lastHash].nElem):
                break
            if (
                self.lastSearchString.equalsIgnoreCase(self.listArray[self.lastHash].str[i])
            ):
                result = self.listArray[self.lastHash].idx[i]
                self.lastFind = i
                break
        return result

    def findNext(self):
        """Begins search in same list as last."""
        # TODO: Check zero indexing.
        result = -1
        # default return
        self.lastFind += 1
        # start with next item in hash list
        if self.lastHash > 0 and self.lastHash <= self.numLists:
            _0 = True
            i = self.lastFind
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < self.listArray[self.lastHash].nElem):
                    break
                if (
                    self.lastSearchString.equalsIgnoreCase(self.listArray[self.lastHash].str[i])
                ):
                    result = self.listArray[self.lastHash].idx[i]
                    self.lastFind = i
                    break
        return result

    def findAbbrev(self, s):
        """Makes a linear search and tests each string until a string is found
        that matches all the characters entered in s.
        """
        result = 0
        if len(s) > 0:
            test1 = s.toLowerCase()
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < self.numElements):
                    break
                test2 = self.stringArray[i][:len(test1)]
                if test1.equalsIgnoreCase(test2):
                    result = i
                    break
        return result

    def get(self, i):
        # TODO: Check zero indexing
        return self.stringArray[i] if i > 0 and i <= self.numElements else ''

    def expand(self, newSize):
        """Expands number of elements.

        Creates a new set of string lists and copies the old strings
        into the new, hashing for the new number of lists.
        """
        # int OldNumLists;
        if newSize > self.numElementsAllocated:
            # OldNumLists = NumLists;
            newStringArray = [None] * newSize
            newNumLists = self.Math.sqrt(newSize)
            elementsPerList = (newSize / newNumLists) + 1
            if newNumLists < 1:
                newNumLists = 1
                # make sure at least one list
            newListArray = [None] * newNumLists
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < self.numLists):
                    break
                # TODO: Check zero indexing.
                # Allocate initial sublists
                newListArray[i].str = [None] * elementsPerList
                newListArray[i].idx = [None] * elementsPerList
                newListArray[i].nAllocated = elementsPerList
                newListArray[i].nElem = 0
            self.numLists = newNumLists
            # has to be set so hash function will work
            # Add elements from old hash list to new hash list
            _1 = True
            i = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    i += 1
                if not (i < self.numElements):
                    break
                # TODO: Check zero indexing
                s = self.stringArray[i]
                hashNum = self.hash(s)
                newListArray[hashNum].nElem += 1
                if newListArray[hashNum].nElem > newListArray[hashNum].nAllocated:
                    self.resizeSubList(newListArray[hashNum])
                newListArray[hashNum].str[newListArray[hashNum].nElem] = s
                newStringArray[self.numElements] = newListArray[hashNum].str[newListArray[hashNum].nElem]
                newListArray[hashNum].idx[newListArray[hashNum].nElem] = i
            # Assign new string and list pointers
            self.stringArray = newStringArray
            self.listArray = newListArray
            self.numElementsAllocated = newSize

    def dumpToFile(self, fname):
        f = File(fname)
        # TODO Auto-generated catch block
        try:
            pw = PrintWriter(f)
            print String.format.format('Number of Hash Lists = %d, Number of Elements = %d', self.numLists, self.numElements)
            print 
            print 'Hash List Distribution'
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < self.numLists):
                    break
                print String.format.format('List = %d, Number of elements = %d', i, self.listArray[i].nElem)
            print 
            _1 = True
            i = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    i += 1
                if not (i < self.numLists):
                    break
                print String.format.format('List = %d, Number of elements = %d', i, self.listArray[i].nElem)
                _2 = True
                j = 0
                while True:
                    if _2 is True:
                        _2 = False
                    else:
                        j += 1
                    if not (j < self.listArray[i].nElem):
                        break
                    print '\"' + self.listArray[i].str[j] + '\"  Idx= ' + self.listArray[i].idx[j]
                print 
            print 'LINEAR LISTING...'
            _3 = True
            i = 0
            while True:
                if _3 is True:
                    _3 = False
                else:
                    i += 1
                if not (i < self.numElements):
                    break
                print i + ' = \"' + self.stringArray[i] + '\"'
            pw.close()
        except FileNotFoundException, e:
            pass # astStmt: [Stmt([]), None]

    def clear(self):
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.numLists):
                break
            self.listArray[i].nElem = 0
            _1 = True
            j = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    j += 1
                if not (j < self.listArray[i].nAllocated):
                    break
                self.listArray[i].str[j] = ''
        _2 = True
        i = 0
        while True:
            if _2 is True:
                _2 = False
            else:
                i += 1
            if not (i < self.numElementsAllocated):
                break
            self.stringArray[i] = ''
        self.numElements = 0
